### PR TITLE
Implemented simple async

### PR DESCRIPTION
**Pull request by timyhac**
Added examples
Modified synchonous read/write to throw exceptions on error

### CODE REVIEW

File: src/libplctag/Tag.cs [hunk 1]
```diff
@@ -10,6 +12,8 @@ namespace libplctag
     public sealed class Tag : IDisposable, ITag
     {
 
+        private const int asyncStatusPollInterval = 2;
```
Comment (jkoplo) [inline src/libplctag/Tag.cs:15]: It's a style thing, but I'm used to constants being named in all caps -> ASYNC_STATUS_POLL_INTERVAL

File: src/libplctag/Tag.cs [hunk 2]
```diff
@@ -111,9 +115,83 @@ private static string GetAttributeString(Protocol protocol, IPAddress gateway, s
...
+            var status = (Status)plctag.read(pointer, 0);
+
+            if (status == Status.Ok)
+                return;
+            else if (status < 0)
+                throw new LibPlcTagException(status);
+
+            do
+            {
+                token.ThrowIfCancellationRequested();
```
Comment (jkoplo) [inline src/libplctag/Tag.cs:144]: Do we want to throw the task cancelled exception to the caller? I could go either way. If you call an async and then cancel it, I think it's okay for it to just return and not throw an exception. In this case we could try/catch the cancellation exception internally or we could add `toke.IsCancelled` to the while condition.

File: src/libplctag/Tag.cs [hunk 3]
```diff
@@ -111,9 +115,83 @@ private static string GetAttributeString(Protocol protocol, IPAddress gateway, s
...
+            var status = (Status)plctag.write(pointer, 0);
+
+            if (status == Status.Ok)
+                return;
+            else if (status < 0)
+                throw new LibPlcTagException(status);
+
+            do
+            {
+                token.ThrowIfCancellationRequested();
```
Comment (jkoplo) [inline src/libplctag/Tag.cs:183]: Same comment on this throw as above.

**Comment by timyhac** [comment]
When developing these I've been referencing the books  "Async in C# 5" and "Concurrency in C#".

From "Concurrency in C#"

Cancellation is treated as a special kind of error. The convention is that canceled code will throw an exception of type OperationCanceledException (or a derived type, such as TaskCanceledException). This way the calling code knows that the cancellation was observed."


Another reason is that the cancellation might not come from the direct caller, so it's helpful for the direct caller to know if the operation completed or was cancelled.


That's a good point about passing a timeout as well, i think so - HttpClient has a timeout property that is used for all requests, what do you think about that behaviour?

**Comment by timyhac** [comment]
What do you mean "do a Task.Run on the synchronous call" - which synchronous call?

File: src/libplctag/Tag.cs (hunk 1 above)
Comment (timyhac) [reply src/libplctag/Tag.cs]: Yep no probs - updated.

**Comment by jkoplo** [comment]
I'm not about to argue with Cleary about async... In retrospect I think it's because our internal code is very brownfield and so I'm used to catching the cancelled exceptions.

I do think we should pass a timeout. You could make the user do it with a linked token or something but that seems obscure. As an end user I would expect a network call to timeout on it's own and not wait forever.

I was thinking of doing `Task.Run(() => plctag.read(pointer, 10))` and then if it timeout start a loop with `Task.Delay`. It's a silly optimization and way premature. It would trade more thread usage off for a msec of potential unnecessary delay. Forget I mentioned it.

**Comment by timyhac** [comment]
Ah yes I see!

Stephen Cleary does have some opinions related to this: https://blog.stephencleary.com/2013/11/taskrun-etiquette-examples-dont-use.html

This whole thing has been very enlightening for me.

I'll update the branch to accept an overload with a timeout as well.

File: src/libplctag/Tag.cs [hunk 4]
```diff
@@ -168,6 +177,15 @@ public void Write(int millisecondTimeout)
 
         }
 
+        public async Task WriteAsync(int millisecondTimeout, CancellationToken token = default)
+        {
+            using (var tcs = CancellationTokenSource.CreateLinkedTokenSource(token))
```
Comment (jkoplo) [inline src/libplctag/Tag.cs:186]: Should this and the one above be `cts`?
Also, what does `CreateLinkedTokenSource` do with a `null`? It seems like that might throw.

File: src/libplctag/Tag.cs [hunk 5]
```diff
@@ -150,7 +150,11 @@ public async Task ReadAsync(CancellationToken token = default)
 
             do
             {
-                token.ThrowIfCancellationRequested();
+                if(token.IsCancellationRequested)
+                {
+                    Abort();
+                    token.ThrowIfCancellationRequested();
+                }
                 await Task.Delay(ASYNC_STATUS_POLL_INTERVAL);
```
Comment (jkoplo) [inline src/libplctag/Tag.cs:158]: I know it's only a 2 msec poll interval, but we should probably stick the token in the `Delay` as long as we have it...

**Comment by jkoplo** [comment]
Oh man, I read that Cleary article quite awhile ago before I had gotten very deep into async. I also hadn't done any ASP.NET at the time. It makes a *ton* more sense to me now! At the time I remember being like, "um, okay... whatever that means".

File: src/libplctag/Tag.cs (hunk 4 above)
Comment (timyhac) [reply src/libplctag/Tag.cs]: Nice catch(es)!

I didn't know this till you asked the question, but the default CancellationToken is CancellationToken.None which works OK with `CreateLinkedTokenSource`

File: src/libplctag/Tag.cs (hunk 5 above)
Comment (timyhac) [reply src/libplctag/Tag.cs]: Done

**Comment by timyhac** [comment]
Thanks for your comments @jkoplo, it was good to verify the behaviour of some of the things I just assumed - let me know if you think anything else is required for this change.

**Comment by jkoplo** [comment]
Nope, looks good.

**Pull request merged**